lib/auth/webauthncli/fido2.go

```diff
@@ -562,54 +582,277 @@ func runOnFIDO2Devices(
+	var info *libfido2.DeviceInfo
+	isFIDO2, err := dev.IsFIDO2()
+	if err != nil {
+		return trace.Wrap(&nonInteractiveError{err: err})
+	}
+	if isFIDO2 {
+		info, err = devInfo(path, dev)
+		if err != nil {
+			return trace.Wrap(&nonInteractiveError{err: err})
+		}
+		log.Debugf("FIDO2: Device %v: info %#v", path, info)
+	} else {
+		log.Debugf("FIDO2: Device %v: not a FIDO2 device", path)
+	}
+	di := makeDevInfo(path, info, isFIDO2)
+
+	// Apply initial filters, waiting for confirmation if the filter fails before
+	// relaying the error.
+	if err := filter(dev, di); err != nil {
+		log.Debugf("FIDO2: Device %v filtered, err=%v", path, err)
+
+		// If the device is chosen then treat the error as interactive.
+		if waitErr := waitForTouch(dev); errors.Is(waitErr, libfido2.ErrNoCredentials) {
+			cancelAll(dev)
+
+			// Escalate error to ErrUsingNonRegisteredDevice, if appropriate, so we
+			// send a better message to the user.
+			if errors.Is(err, errNoRegisteredCredentials) {
+				err = ErrUsingNonRegisteredDevice
+			}
+
+		} else {
+			err = &nonInteractiveError{err: err}
+		}
 		return trace.Wrap(err)
 	}
 
-	// List/select devices.
+	// Run the callback.
 	cb := withPINHandler(withRetries(deviceCallback))
-	dev, requiresPIN, err := findAndSelectDevice(ctx, filter, cb)
+	requiresPIN, err := cb(dev, di, "" /* pin */)
+	log.Debugf("FIDO2: Device %v: callback returned, requiresPIN=%v, err=%v", path, requiresPIN, err)
 	if err != nil {
 		return trace.Wrap(err)
 	}
 
-	if err := ackTouch(); err != nil {
+	if err := firstTouchAck(); err != nil {
 		return trace.Wrap(err)
 	}
 
+	// Cancel other devices only on success. This avoids multiple cancel attempts
+	// as non-chosen devices return FIDO_ERR_KEEPALIVE_CANCEL.
+	cancelAll(dev)
+
 	if !requiresPIN {
 		return nil
 	}
 
-	// Selected device requires PIN, let's use the prompt and run the callback
-	// again.
-	pin, err := prompt.PromptPIN()
+	// Ask for PIN, prompt for next touch.
+	pin, err := pinPrompt.PromptPIN()
 	switch {
 	case err != nil:
 		return trace.Wrap(err)
 	case pin == "":
 		return libfido2.ErrPinRequired
 	}
-
-	// Prompt a second touch after reading the PIN.
-	ackTouch, err = prompt.PromptTouch()
+	ackTouch, err := pinPrompt.PromptTouch()
 	if err != nil {
 		return trace.Wrap(err)
 	}
 
-	// Run the callback again with the informed PIN.
-	// selectDevice is used since it correctly deals with cancellation.
 	cb = withoutPINHandler(withRetries(deviceCallback))
-	_, err = selectDevice(ctx, pin, dev, cb)
-	if err != nil {
+	if _, err := cb(dev, di, pin); err != nil {
 		return trace.Wrap(err)
 	}
-
 	return trace.Wrap(ackTouch())
 }
 
+func devInfo(path string, dev FIDODevice) (*libfido2.DeviceInfo, error) {
+	const infoAttempts = 3
+	var lastErr error
+	for i := 0; i < infoAttempts; i++ {
+		info, err := dev.Info()
+		if err == nil {
+			return info, nil
+		}
+
+		lastErr = err
+		log.Debugf("FIDO2: Device %v: Info failed, retrying after interval: %v", path, err)
```

Comment:
```suggestion
		log.Debugf("FIDO2: Device %v: Info failed, retrying after interval: %v", path, fido2RetryInterval)
```

Reply:
probably should include the err as well, but I couldn't tell if you intended to

Reply:
It was written to log the error but not the interval, so yeah it was intentional.

Changed to:

```diff
- log.Debugf("FIDO2: Device %v: Info failed, retrying after interval: %v", path, err)
+ log.Debugf("FIDO2: Device %v: Info failed, retrying after %s: %v", path, fido2RetryInterval, err)
```